examples/step-86/step-86.cc

```diff
@@ -0,0 +1,1138 @@
+  // discussion in step-77). We could in principle not provide this function to
+  // the time stepper, and instead select a specific solver on the command line
+  // by using the `-ksp_*` options of PETSc. However, by providing this
+  // function, we can use a specific solver and preconditioner for the linear
+  // system, and still have the possibility to change them on the command line.
+  //
+  // Providing a specific solver is more in line with the way we usually do
+  // things in other deal.II examples, while letting PETSc choose a generic
+  // solver, and changing it on the command line via `-ksp_type` is more in line
+  // with the way PETSc is usually used. Both options are available here, since
+  // we can still change both the solver and the preconditioner on the command
+  // line via `-user_ksp_type` and `-user_pc_type` options.
+  //
+  // In any case, recall that the Jacobian we built in the previous function is
+  // always of the form
+  // @f[
+  //   J_\alpha = \alpha M + A
+  // @f]
+  // where $M$ is a mass matrix and $A$ a Laplace matrix. $M$ is symmetric and
+  // positive definite; $A$ is symmetric and at least positive semidefinite;
+  // $\alpha> 0$. As a consequence, the Jacobian matrix is a symmetric and
+  // positive definite matrix, which we can efficiently solve with the Conjugate
+  // Gradient method, along with either SSOR or (if available) the algebraic
+  // multigrid implementation provided by PETSc (via the Hypre package) as
+  // preconditioner.
+  template <int dim>
+  void
+  HeatEquation<dim>::solve_with_jacobian(const PETScWrappers::MPI::Vector &src,
+                                         PETScWrappers::MPI::Vector       &dst)
+  {
+    TimerOutput::Scope t(computing_timer, "solve with Jacobian");
+
+#if defined(PETSC_HAVE_HYPRE)
+    PETScWrappers::PreconditionBoomerAMG preconditioner;
+    preconditioner.initialize(jacobian_matrix);
+#else
+    PETScWrappers::PreconditionSSOR preconditioner;
+    preconditioner.initialize(
+      jacobian_matrix, PETScWrappers::PreconditionSSOR::AdditionalData(1.0));
+#endif
+
+    SolverControl           solver_control(1000, 1e-8 * src.l2_norm());
+    PETScWrappers::SolverCG cg(solver_control);
+    cg.set_prefix("user_");
+
+    cg.solve(jacobian_matrix, dst, src, preconditioner);
+
+    pcout << "     " << solver_control.last_step() << " linear iterations."
+          << std::endl;
+  }
+
+
+  // @sect4{The HeatEquation::prepare_for_coarsening_and_refinement() function}
+  //
+  // The next block of functions deals with mesh refinement. We split this
+  // process up into a "decide whether and what you want to refine" and a
+  // "please transfer these vectors from old to new mesh" phase, where the first
+  // also deals with marking cells for refinement. (The decision whether or
+  // not to refine is done in the lambda function that calls the current
+  // function.)
+  //
+  // Breaking things into a "mark cells" function and into a "execute mesh
```

Comment:
This is because you are not considering the point of view of the solver, and you are not discussing the actual callback, the `decide_for_coarsening_and_refinement`. For me, it is natural to first take a decision, inform the solver about that, and then the solver can do what it needs to be done before performing the actual remesh + transfer

Reply:
I was thinking about why you structured it this way. I *think* the simplest way would have been if the `decide_...` function really only does that: *Decide* whether we want to do mesh refinement. Then the marking along with the actual refinement would happen in the second callback, which would also be natural (keep marking and refinement together). But I suspect that you didn't do it that way because you don't have the solution vector available in the second callback (you only have a whole bunch of vectors), and so the marking needs to happen in the first callback.

I don't know whether it would be possible to query the TS object for the current solution vector? If that were the case, we could provide that in the second callback.

Reply:
This is the relevant PETSc code https://gitlab.com/petsc/petsc/-/blob/main/src/ts/interface/ts.c?ref_type=heads#L3841. `resizesetup` runs the deal.ii `decide` callback (actually, I wrote this in PETSc to support this from deal.II). The solution vector is the first vector of the list when doing the transfer.  It looked clean to me, but if you think a different approach would be cleaner, please change it.